// cx/dist/hooks.js
import { getCurrentInstance } from "cx/ui";
import { getSelector, computable, Ref } from "cx/data";
import { isObject, isArray } from "cx/util";

function useStore() {
	return getCurrentInstance().store;
}
function useStoreMethods() {
	return getCurrentInstance().store.getMethods();
}
function ref(info) {
	if (isObject(info)) {
		if (info.bind) return useStore().ref(info.bind, info.defaultValue);
		if (info.get) return info;
		if (info.expr) throw new Error("Not supported yet.");
	}

	return getSelector(info);
}

function useEffect(callback) {
	var destroyCallback = callback();
	if (destroyCallback) getCurrentInstance().subscribeOnDestroy(destroyCallback);
}
function useCleanup(cleanupCallback) {
	var unsubscribe = getCurrentInstance().subscribeOnDestroy(cleanupCallback);
	return function() {
		unsubscribe();
		cleanupCallback();
	};
}

function useInterval(callback, timeout) {
	var timer = setInterval(callback, timeout);
	return useCleanup(function() {
		clearInterval(timer);
	});
}

function addExploreCallback(callback) {
	var instance = getCurrentInstance();
	if (!instance.computables) instance.computables = [];
	instance.computables.push(callback);
	return function() {
		instance.computables = instance.computables.filter(function(cb) {
			return cb !== callback;
		});
	};
}
function useTrigger(args, callback) {
	if (!isArray(args)) throw new Error("First argument to addTrigger should be an array.");
	var selector = computable.apply(void 0, args.concat([callback])).memoize();
	return addExploreCallback(selector);
}

var key = 0;
function useState(defaultValue) {
	var _instance$setState;

	var instance = getCurrentInstance();
	var storeKey = "_state" + ++key;
	instance.setState(((_instance$setState = {}), (_instance$setState[storeKey] = defaultValue), _instance$setState));
	return new Ref({
		get: function get() {
			return instance.state[storeKey];
		},
		set: function set(value) {
			var _instance$setState2;

			return instance.setState(
				((_instance$setState2 = {}), (_instance$setState2[storeKey] = value), _instance$setState2)
			);
		}
	});
}

function createLoacalStorageRef(key) {
	var store = useStore();
	return new Ref({
		get: function get() {
			var json = localStorage.getItem(key);
			if (!json) return localStorage.hasOwnProperty(key) ? null : undefined;
			return JSON.parse(json);
		},
		set: function set(value) {
			if (value === undefined) localStorage.removeItem(key);
			else localStorage.setItem(key, JSON.stringify(value));
			store.meta.version++;
			store.notify();
		}
	});
}

export {
	addExploreCallback,
	createLoacalStorageRef,
	ref,
	useCleanup,
	useEffect,
	useInterval,
	useState,
	useStore,
	useStoreMethods,
	useTrigger
};
