// @mui-treasury/layout/components/Content.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import cx from 'clsx';
import { makeStyles } from '@material-ui/styles';
import { useTransitionStyles } from '../styles';
import { useLayoutCtx } from '../hooks';
const useStyles = makeStyles(() => ({
  root: {
    flexGrow: 1
  }
}));

const Content = (_ref) => {
  let {
    component: Component,
    className,
    children,
    style
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, ["component", "className", "children", "style"]);

  const ctx = useLayoutCtx();
  const {
    getSidebarGap,
    getWidth,
    content
  } = ctx;
  const styles = useStyles(props);
  const transitionStyles = useTransitionStyles();
  return React.createElement(React.Fragment, null, React.createElement(Component, _extends({}, props, {
    className: cx(styles.root, transitionStyles.root, className),
    style: _extends({}, style, {
      marginLeft: getSidebarGap(content),
      width: getWidth(content)
    })
  }), typeof children === 'function' ? children(ctx) : children));
};

Content.propTypes = {
  className: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
  component: PropTypes.elementType,
  style: PropTypes.shape({})
};
Content.defaultProps = {
  className: undefined,
  component: 'main',
  style: undefined
};
export default Content;
